### DISCUSSION TITLE
Download the image

### DISCUSSION THREAD

**Discussion by deadcoder0904** [neg4n/next-api-og-image#10] [Q&A] [answered]
I've been trying to get it to work but having unfortunate errors.

My `api/og-image.ts` looks like:

```ts
import React from 'react'
import ReactDOM from 'react-dom/server'
import { withOGImage } from 'next-api-og-image'

import OGImagePreview from '@/pages/og-image-preview'
import { OGImage, Author } from '@/types/index'

const html = async ({
	title,
	description,
	date,
	readingTime,
	slug,
	authors,
}: Record<keyof OGImage, string | string[] | any[]>): Promise<string> => {
	const el = React.createElement(OGImagePreview, {
		title: title as string,
		description: description as string,
		date: date as string,
		readingTime: readingTime as string,
		slug: slug as string,
		authors: authors as Author[],
	})

	const body = ReactDOM.renderToStaticMarkup(el)

	const baseCSS = `
		.font-alegreya {
			font-family: Alegreya;
		}
		.font-jetbrains {
			font-family: 'JetBrains Mono';
		}
	`

	return `
		<html>
			<head>
				<meta charset="utf-8" />
				<link rel="preconnect" href="https://fonts.googleapis.com" />
				<link rel="preconnect" href="https://fonts.gstatic.com" crossOrigin="true" />
				<link
					href="https://fonts.googleapis.com/css2?family=Alegreya:wght@800&display=swap"
					rel="stylesheet"
				/>
				<link
					href="https://fonts.googleapis.com/css2?family=JetBrains+Mono&display=swap"
					rel="stylesheet"
				/>
				<link href="https://unpkg.com/tailwindcss@^2/dist/tailwind.min.css" rel="stylesheet" />
				<style>${baseCSS}</style>
			</head>
			<body>
				${body}
			</body>
			<script src="https://unpkg.com/tailwindcss-jit-cdn"></script>
		</html>
	`
}

const nextApiOgImageConfig = {
	// 'Content-Type' HTTP header
	contentType: 'image/png',
	// 'Cache-Control' HTTP header
	cacheControl: 'max-age 3600, must-revalidate',
	// NOTE: Options within 'dev' object works only when process.env.NODE_ENV === 'development'
	dev: {
		// Whether to replace binary data (image/screenshot) with HTML
		// that can be debugged in Developer Tools
		inspectHtml: true,
	},
	html,
}

export default withOGImage<keyof OGImage>(nextApiOgImageConfig)
```

How do I download the image itself?

I was trying to use simple `fetch` & also `ky` but not sure how to download it:

```ts
import ky from 'ky'

import {
	getAllEssays,
	getAllTutorials,
	getAllSnippets,
	getAllTips,
	getEssayBySlug,
	getTutorialBySlug,
	getSnippetBySlug,
	getTipBySlug,
} from '@/utils/mdx'
import { siteMetadata } from '@/_data/index'

const { siteUrl } = siteMetadata

export const generateOGImage = () => {
	try {
		const essays = getAllEssays()
		const tutorials = getAllTutorials()
		const snippets = getAllSnippets()
		const tips = getAllTips()

		essays.forEach(async ({ slug }) => {
			const { meta, code } = await getEssayBySlug(slug)
			const { title, description = '', date, readingTime, authors } = meta
			let data = new URLSearchParams()
			data.append('title', title)
			data.append('description', description)
			data.append('date', JSON.stringify(date))
			data.append('readingTime', JSON.stringify(readingTime))
			data.append('slug', slug)
			data.append('authors', JSON.stringify(authors))

			const options = {
				method: `POST`,
				body: data,
			}

			fetch(`${siteUrl}/api/og-image`, options)
				.then((data) => {
					if (!data.ok) throw data

					console.log(data.body)
				})
				.catch((error) => {
					console.error(error)
				})

			const json = await ky
				.post('api/og-image', {
					prefixUrl: siteUrl,
					json: {
						title: meta.title,
						description: meta.description,
						date: meta.date,
						readingTime: meta.readingTime,
					},
				})
				.json()
			console.log(json)
		})
	} catch (e) {
		console.error(e)
	} finally {
	}
}
```

I want the image from this API to be downloaded. How do I do it?

**Comment by neg4n** [comment, marked as answer]
You can try something like that:

In api route (`pages/api/ogimage.js`):

```js
import { withOGImage } from "next-api-og-image";

export default withOGImage({
  template: { html: ({ test }) => `<h1>${test}</h1>` },
  dev: { inspectHtml: false },
});
```

then in regular Next.js page (`pages/test.js`):

```js
import fetch from "isomorphic-unfetch";

export default function TestPage({ image }) {
  const src = `data:image/png;base64,${image}`;
  return (
    <div>
      <img src={src} />
    </div>
  );
}

export async function getStaticProps() {
  const response = await fetch(
    "http://localhost:3000/api/ogimage?test=populated%20at%20build%20time"
  );

  const arrayBuffer = await response.arrayBuffer();
  const base64 = Buffer.from(arrayBuffer, "base64").toString("base64");

  return {
    props: {
      image: base64,
    },
  };
}
```

Navigating to `http://localhost:3000/test` should have <img> tag with image generated by the library. And it will be generated during the build-time! Hope that helps @deadcoder0904  =) Don't forget to mark discussion as answered if you manage to get your issue resolved! Cheers

**Reply by deadcoder0904** [reply]
Hey @neg4n, I don't want to `fetch` it in a `page` like that. I want to do it in a separate function so I can save it to the post folder directly. Here's what I have:

```ts
import fs from 'fs'
import path from 'path'
import fetch from 'isomorphic-unfetch'
import FileType from 'file-type'

import {
	getAllEssays,
	getEssayBySlug,
} from '@/utils/mdx'

import { authors } from '@/_data/index'

const saveImageFromAPI = async (url: string, folderPath: string) => {
	const response = await fetch(url)
	const arrayBuffer = await response.arrayBuffer()
	const buffer = Buffer.from(arrayBuffer)
	const fileType = await FileType.fromBuffer(buffer)

	console.log(fileType)

	if (fileType?.ext) {
		const outputFileName = `og-image.${fileType.ext}`
		try {
			fs.createWriteStream(path.join(folderPath, outputFileName)).write(buffer)
		} catch (e) {
			console.error(e)
		}
	} else {
		console.log(
			'File type could not be reliably determined! The binary data may be malformed! No file saved!'
		)
	}
}

export const generateOGImage = () => {
	try {
		const essays = getAllEssays()

		essays.forEach(async ({ slug }) => {
			const { meta, folderPath } = await getEssayBySlug(slug)
			const { title, description = '', date, readingTime, authors: metaAuthors } = meta

			const postAuthors = authors.filter(({ name }) => metaAuthors.includes(name))
			const params = {
				title,
				description,
				date,
				readingTime,
				authors: postAuthors,
			}
			const searchParams = Object.keys(params)
				.map((k) => {
					return `${encodeURIComponent(k)}=${encodeURIComponent(
						JSON.stringify(params[k as keyof typeof params])
					)}`
				})
				.join('&')

			saveImageFromAPI(`http://localhost:3000/api/og-image?${searchParams}`, folderPath)
		})
	} catch (e) {
		console.error(e)
	} finally {
	}
}
```

I loop over `essays` & try to get `title`, `date`, etc... & then save it to file but see the `console.log`.

I get `undefined` on `fileType` so it can't reliably detect `buffer`. I've looked at https://chrisfrew.in/blog/saving-images-in-node-js-using-fetch-with-array-buffer-and-buffer/ & did the exact same thing but doesn't work.

I have probably looked everywhere using different libraries like `ky`, `axios`, `got` but still have not manage to solve this problem.

What is the return type? I would love to solve it but really unable to. I managed to get the file downloaded but it is corrupt & never opens :(

**Reply by deadcoder0904** [reply]
Weird, I just made a repro using `got` & it worked → https://github.com/deadcoder0904/got-image-buffer-to-disk

**Reply by deadcoder0904** [reply]
@neg4n it's not working in my main project. i'm assuming it's because i have other types like `Date` & `IReadTimeResults` (from `reading-time`)

is there any way to pass params without using query params? i haven't done node in a while but i'm assuming `req.body` should work, right? also, what about sending it as a `json`? or `.post()`? would love to know using `got` module since it works fine with another project just fine!

I just pushed [a new commit](https://github.com/deadcoder0904/got-image-buffer-to-disk/commit/c1aed7842b337d4f8b68c379550035a5537a20a5) that uses `Date` type & it doesn't work. I used it with `got.post` while passing `json` parameter. Would love to make this work? I used a `post` request rather than `get` because I can't pass `json` any other way.

**Reply by neg4n** [reply]
@deadcoder0904 You can pass these props in query, just pass them as strings since query param is always `Record<string, string>` or `Record<string, string[]>` but the second option is very unlikely to occur in real world scenario and probably will be removed in the future. 😇 

By the way, is it really good approach to output images to the directory? I think the only usage is to either statically/dynamically generate open-graph images in serverless environment and populate them in pages or *(not tested)* generate them during webpack bundling (here output to files seems reasonable)

**Reply by deadcoder0904** [reply]
> @deadcoder0904 You can pass these props in query, just pass them as strings since query param is always `Record<string, string>` or `Record<string, string[]>` but the second option is very unlikely to occur in real world scenario and probably will be removed in the future. 😇

I tried it using `JSON.stringify` but it didn't work at all. I also have `IReadTimeResults` type which didn't work either. I think I probably need `req.body` rather than passing it a query parameter using `req.query` so I can pass `json`.

> By the way, is it really good approach to output images to the directory? I think the only usage is to either statically/dynamically generate open-graph images in serverless environment and populate them in pages or _(not tested)_ generate them during webpack bundling (here output to files seems reasonable)

Idk. I just wanted a way to store the images in the blog itself so I wanted to dynamically generate it & store it in each posts folder.

Anyways would love to know a solution? I can't manage to get this work & I've probably spent many days on it now.

**Reply by neg4n** [reply]
@deadcoder0904 Sorry for the late response but i really had no time :(
> I tried it using JSON.stringify but it didn't work at all. I also have IReadTimeResults type which didn't work either. I think I probably need req.body rather than passing it a query parameter using req.query so I can pass json.

You must not stringify your query but cast __every single value separately__ to string.
The thing with passing values as body seems like a good idea. I will implement it soon 😇

**Reply by deadcoder0904** [reply]
@neg4n nice `req.body` seems like a cool option.

so do you mean doing `as string`? i've done that but it can't work for `IReadTimeResults` or `Date` as the first one is a json & for second one I tried `.toString()` but still errors haha.

And yes if you do have time, can you make my example work → https://github.com/deadcoder0904/got-image-buffer-to-disk?

I'll then try to reimplement it in my blog project & let you know if it works here :)

**Reply by neg4n** [reply]
> so do you mean doing as string? i've done that but it can't work for IReadTimeResults or Date as the first one is a json & for second one I tried .toString() but still errors haha.

> You must not stringify your query but cast every single value separately to string.

Code from [deadcoder0904/got-image-buffer-to-disk](https://github.com/deadcoder0904/got-image-buffer-to-disk)

`lib/generate-og-image.ts`

```ts
import fs from "fs";
import path from "path";
import got from "got";

const saveImageFromAPI = async (
  url: string,
  params: Record<string, string>,
  folderPath: string
) => {
  try {
    const resp = await got
      .post(url, {
        searchParams: {
          ...params,
        },
      })
      .buffer();
    fs.writeFileSync(path.join(folderPath, "og-image.jpeg"), resp);
  } catch (e) {
    console.error(e);
  }
};

export const generateOGImage = () => {
  try {
    const folderPath = "public/";

    const params = {
      title: "Hello OG Image",
      date: new Date(),
      someComplexData: {
        yeah: "super",
        thats: "epic",
        foo: {
          bar: 10,
        },
      },
    };

    const API_URL = `http://localhost:3000/api/og-image`;

    saveImageFromAPI(
      API_URL,
      {
        title: params.title,
        date: params.date.toString(),
        someComplexDataYeah: params.someComplexData.yeah,
        someComplexDataThats: params.someComplexData.thats,
        someComplexDataFooBar: `${params.someComplexData.foo.bar}`,
      },
      folderPath
    );
  } catch (e) {
    console.error(e);
  } finally {
  }
};
```

`pages/og-image.ts`
```ts
import { withOGImage } from "next-api-og-image";

export default withOGImage({
  html: ({
    title,
    date,
    someComplexDataYeah,
    someComplexDataThats,
    someComplexDataFooBar,
  }) =>
    `<h1>${title} ${date} ${someComplexDataYeah}} ${someComplexDataThats}${someComplexDataFooBar}</h1>`,
  dev: { inspectHtml: false },
});
```

if you want strongly typed query:

https://user-images.githubusercontent.com/57688858/137306170-13499126-dd64-4be3-9883-9f9a14ea88c3.mov

# Result

![og-image](https://user-images.githubusercontent.com/57688858/137306429-f260fbda-e763-4f16-abb6-d61c7868598a.jpeg)

@deadcoder0904

**Reply by neg4n** [reply]
Also @deadcoder0904, make sure to check out https://github.com/neg4n/next-api-og-image/releases/tag/2.0.0 🤟🏻

**Reply by deadcoder0904** [reply]
@neg4n the example you've given does not pass an `object`, it instead passes a `string` everytime while destructuring it like `someComplexDataYeah`, `someComplexDataThats` & `someComplexDataFooBar`.

I have an authors object which I want to pass & the authors has 2 properties: `name` & `twitter`. I want to pass that completely instead of individually as I have to do lots of string conversions which are not ideal. I think this is due to query params so #11 should probably fix it.

Here's my latest commit → https://github.com/deadcoder0904/got-image-buffer-to-disk/commit/afa547195fe8172e6eb439473fbae71a66f72717

See how I need to do a bunch of conversions & it still doesn't work for `Date` or `authors` field but yeah it does save to a file now.

So all I need is a way to send `json` or basically objects. That would make this project perfect :)

**Reply by neg4n** [reply]
I'll implement option to pass properties in body soon. Also, I encourage you @deadcoder0904 to do some pull requests 😃 This discussion will be marked as answered since original question/problem has already been resolved. Cheers
